### DISCUSSION TITLE
How to enable copying from the content?

### DISCUSSION THREAD

**Discussion by faressoft** [LaunchMenu/LaunchMenu#192] [Q&A] [answered]
How to enable the default OS shortcut to copy the selected content (e.g. Command+C)?

<img width="868" alt="Screenshot 2023-09-20 at 02 34 22" src="https://github.com/LaunchMenu/LaunchMenu/assets/1908658/b7acf38b-0fbe-4095-a8fa-30293ba4a407">

**Comment by TarVK** [comment]
I believe this should be what you're looking for: [common-actions#copy-and-paste](https://launchmenu.github.io/docs/concepts/interaction/common-actions#copy-and-paste)
Let me know if this doesn't answer your question!

**Comment by faressoft** [comment]
@TarVK 

Thanks for your response.

`copyTextHandler` doesn't seem what I'm looking for. `copyTextHandler` takes a string and copies it. 

What I am trying to achieve is to make the content part copyable. The screenshot I provided shows a selected text, and I want to copy what I have selected.

The screenshot may be confusing because of the green color, but I simply changed the default color using:

```
::selection {
  color: black;
  background: #BCFF1F;
}
```

**Reply by TarVK** [reply, marked as answer]
Aaah right, that's a more interesting question. LM was kind of designed with the idea that you shouldn't ever need to use mouse text selection at all, copying should be done through the context-menu. So because of this idea, we've not made it easy to do this sort of stuff.

However, I think the following code is the source of your issue: [KeyHandler.ts#L139](https://github.com/LaunchMenu/LaunchMenu/blob/development/packages/core/src/keyHandler/KeyHandler.ts#L139).
So you could modify that code. It would probably be better to create your own ctrl+c handler on your content, and when receiving this event delete the `event.original` entry so this code won't get called. That would reduce other parts of the application possibly being negatively affected by something that would otherwise be a rather fundamental change.

**Reply by faressoft** [reply]
@TarVK 

Thank you. I managed to enable copying by wrapping component with:

```typescript
export function copyable<T>(WrappedComponent: new (props: T) => Component<T, any>) {
  return class extends Component<T> {
    componentDidMount() {
      document.addEventListener('keydown', this.handleKeyDown);
    }

    componentWillUnmount() {
      document.removeEventListener('keydown', this.handleKeyDown);
    }

    handleKeyDown = (event: KeyboardEvent) => {
      if ((event.metaKey || event.ctrlKey) && event.key === 'c') {
        document.execCommand('copy');
      }
    };

    render() {
      return <WrappedComponent {...this.props} />;
    }
  };
}
```

**Reply by TarVK** [reply]
Oh great! That's an awesome solution!
Only remaining tiny improvement I see, is using a key-pattern from the settings to compare against. 
But ctrl/meta+c is universal enough that it probably doesn't matter a great deal.
